### ISSUE TITLE
feat(firebaseRef): consider exposing firebase events as an Observable

### ISSUE THREAD

**Issue by robwormald** [angular/angularfire2#55]
Scenario - i'd like to use AngularFire/Firebase in a redux/rxjs style app, where i might mix and match data from various sources. Rather than using a FirebaseObservable directly, I might want to get an Observable of the raw events emitting by a firebaseRef. ('child_added' etc)

**Comment by jeffbcross** [comment]
Thanks @robwormald. Would you want to have a single Observable for all events for a ref? Would an API like the following make you happy?

```typescript
constructor(af:AngularFire) {
  af.listen('/questions').subscribe(e => {
    switch (e.type) {
      case 'child_added':
        console.log('added', e.snapshot.val(), e.prevKey);
        break;
      default:
        console.log('unknown event', e.type);
    }
  })
}
```

Default could be to listen to all events, with an options object to specify either include or exclude events:
`af.listen('/questions', { exclude: 'value', query: ... })`

That brings up another question, would you want the option to suppress `child_added` events for the initial value like we are planning to do with `list()` or just leave it up to the reducer to handle such complexity? I'd lean toward leaving heavy lifting to the reducer so this API can be simple.

**Comment by TylerEich** [comment]
Maybe something like `af.listen( path[, eventArray ] )`, e.g. `af.listen( '/questions' )` will listen for all events, `af.listen( '/questions', [ 'value', 'child_removed' ] )` only listens for the `value` and `child_added` events.

**Comment by robwormald** [comment]
@jeffbcross yeah i think that API would be fine. i think it would be fine to leave the downstream consumer to deal with filtering out events.

**Comment by robwormald** [comment]
and i like @TylerEich's API if it doesn't add too much complexity.

**Comment by davideast** [comment]
I like this idea, but it only makes sense to listen to child events together. There's really no need to use a `value` event with a `child_added` event. And the `FirebaseListObservable` is all of the child events munged into one array.

If you want granularity on the events, we could pass an event object back in `.list()`, similar to what @jeffbcross proposed.

Otherwise we could create a `ChildEventObservable` that can take in an array of child events.

Thoughts @robwormald and @TylerEich?

**Comment by TylerEich** [comment]
In my opinion, the `FirebaseListObservable` should have one responsibility: taking the raw object from `value` events, distilling them into arrays, and emitting the result to subscribers. I don't think the ability to listen to events belongs in `FirebaseListObservable`. The same applies for `FirebaseObjectObservable`.

Similarly, I don't think there should be an event object parameter for `.list()`. Listening for specific events should be wholly separate from the `.list()` or `.object()` methods.

+1 for `ChildEventObservable`.

I'm still in favor of a `ref.listen()` method. I think it ought to return a `ChildEventObservable`.

Benefits of this approach:
* If a person just wants a simple array that stays synced with Firebase, `ref.list()` is all they need. For an object, it's `ref.object()`. Super simple, fits well into unidirectional data flow.
* If a person wants more granular control, they can execute `ref.listen('path', ['events', 'to', 'listen', 'to'])`. They can use Rx reducers to execute their own logic.
* `FirebaseListObservable` and `FirebaseObjectObservable` are nothing but shortcuts for `ref.listen('path', ['value']).map(doTheArrayThing)` and `ref.listen('path', ['value']).map(doTheObjectThing)` respectively, plus some convenience methods like `.add()` and `.remove()`

TL;DR Keep the event management separate from the convenience methods for arrays and objects.

**Comment by lokeshjain2008** [comment]
@davideast What you have decided? when we can expect this to be reality or is there any way to do it?

**Comment by shlomiassaf** [comment]
Is this an option?

```ts
const list = angularfilre.list('https://mysite.firebase.io');
list.subscribe(...) // regular
list.childAdded().subscrube(...)
```

**Comment by TylerEich** [comment]
@shlomiassaf +1 for `subscrube(...)` 😜

But in all seriousness, I still think it would be better to keep the event propagation and synced list stuff separated

**Comment by shlomiassaf** [comment]
@TylerEich you got me there :/

Do it separated, do it RxJS style, make the user `import 'add/something`

But provide the support in one place...

**Comment by TylerEich** [comment]
Agreed. An Rx API would be nice to have built in

**Comment by shlomiassaf** [comment]
@TylerEich It's straight forward, I think they do it in ngrx @robwormald

**Comment by davideast** [comment]
@shlomiassaf @TylerEich @lokeshjain2008 

Right now this is on the back burner. We're working on cleaning up the library with the Firebase 3.0 SDK changes, RC3 support, and new Firebase features like storage.

I do like this idea though. If we can hash out a good solution it would be great addition to the library.

**Comment by davideast** [comment]
Closing due to inactivity. Will open for discussion later.

**Comment by alexjoverm** [comment]
Any update on this? Personally I consider it very important... Right now, the idea of angularfire2 is amazing, but works only for small apps where you bind firebase result to you component directly, since it always return all items, and there is no way to listen for updates or deletes.

Most of times you need to listen only to the added children (after init) and listen for updates, deletes... For example if you use ngrx store.

**Comment by TylerEich** [comment]
I agree with @alexjoverm. I've been scratching my head looking for ways to integrate AngularFire2 and redux architecture. I realize there are ways to do this using the Firebase 3.0 SDK directly, but I love the way AngularFire2 automatically Observable-ifies the events and makes it easier to use arrays in the database.

**Comment by alexjoverm** [comment]
Exactly, for now I could use it getting the firebase ref from angularfire2. It is exactly like using the raw SDK, but at least keeps angular's DI:

```ts
import { FirebaseRef } from 'angularfire2'

@Injectable()
export class FirebaseService {

  private listsRef

  constructor(
    private store: Store<AppState>,
    @Inject(FirebaseRef) private firebase
  ) {
    this.listsRef = firebase.database().ref('lists')
 }

 init () {
    this.listsRef.on('child_added', listSnap => {
       ....
    })
  }

...
```

**Comment by theunreal** [comment]
So.. using `child_added`, `child_changed` and `child_removed` is impossible with Angularfire2 unelss you inject the Firebaseref and mixing all together? 
Those methods are important main core methods of Firebase, which are the base of many applications. Watching for new items, changes and removes is highly important. Apps that need to use this methods cannot use Angularfire2.

**Comment by vitorrd** [comment]
I don't understand this. Wasn't this library created as "The official Angular 2 library for Firebase"? Why doesn't it contain **basic** features such as event support? If one is to hook into Firebase itself, dodging NgFire2, why would one use it in the first place?

**Comment by katowulf** [comment]
There's little value of creating a library that provides exactly the same functionality as an existing SDK. Instead, this library focuses on components needed to use Angular and Firebase together effectively: removing boilerplate, handling nuanced complexities like converting real-time data streams into Observables, and providing some common functionality for the Auth process.

As with any library, balance and focus are necessary. We can't support every use case, edge case, or need. Where possible, we'll avoid duplicating functionality that already exists in the SDK, and focus on providing value by providing more Angular-centric tools, such as observable collections and objects.

Locking this thread as continued discussion of long-past design decisions will be counter productive. Having said this, as with any library, there's always room for improvement. 

Please feel free to file specific feature requests for specific needs, **being sure to note where this library can add value over existing SDK methods** rather than simply replicating them.
